Fix/mcp tool adapter connects to the mcp server in a stateless mode

SongChiYoung opened the pull request:
<!-- Thank you for your contribution! Please review https://microsoft.github.io/autogen/docs/Contribute before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

This PR introduces a reusable, event-driven McpSessionActor component for AutoGen’s MCP tool adapters.
Previously, each tool instance created its own isolated ClientSession, making it difficult to share a single session context across tools. This change solves that by:
 - Introducing a McpSessionActor that manages a single session via an async task with a queue.
 - Allowing safe concurrent usage of the same session by multiple tools.
 - Supporting serialization/deserialization of the session actor for config-based workflows.
 - Ensuring graceful shutdown of the session, even in atexit or test teardown scenarios.

This addresses hanging issues during test teardown and resolves RuntimeError: Attempted to exit cancel scope in a different task errors in anyio.

## Related issue number

Closes #6198

## Checks

- [ ] I've included any doc changes needed for <https://microsoft.github.io/autogen/>. See <https://github.com/microsoft/autogen/blob/main/CONTRIBUTING.md> to build and test documentation locally.
- [x] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [x] I've made sure all auto checks have passed.

SongChiYoung commented:
### Notes on Lifecycle Management and Design Decisions

- This PR does not include a routine to reinitialize the session when an Agent or Team stops for various reasons. Instead, the actor shuts down gracefully when the process terminates.  
  It's unclear whether the `on_reset` method is expected to be invoked in such cases, so I took a conservative approach. I'm happy to adjust this if there's guidance from the maintainers.

- As seen in the test cases, when the session is used outside of an Agent (i.e., called directly), users must explicitly call the `close()` method.  
  This design assumes lazy initialization and does not cause hangs as long as the session is constructed but not used.  
  Supporting `with`-statement awareness was not included in this PR to keep things simple, but it can be added in a separate PR if the community sees value in it.

- Currently, the session is not reset inside `on_reset()` since doing so would require another `initialize()` on the next call, adding more complexity.  
  If needed, I’m open to revisiting this, either in this PR or a follow-up, depending on guidance from domain experts or maintainers.

In summary: there may be edge cases that lead to a hang **only** if the session is used without an Agent and not closed explicitly.  
This is considered a usage-side responsibility. Adding more complexity to guard against such nonstandard usage might lead to over-engineering.  
As for `on_reset`, I tried to follow the “simple is best” philosophy by relying on graceful shutdown rather than adding extra lifecycle management.

If there’s a better idiomatic approach for AutoGen or MCP usage patterns, I’d greatly appreciate your advice.

ekzhu commented on python/packages/autogen-ext/src/autogen_ext/tools/mcp/_session.py:
```diff
@@ -26,3 +31,105 @@ async def create_mcp_server_session(
+            async with create_mcp_server_session(self.server_params) as session:
+                await session.initialize()
+                while True:
+                    cmd = await self._command_queue.get()
```
Do we need a queue for processing the session commands?

SongChiYoung replied:
Same issue from under that.
I will summary it at the under comment.

ekzhu commented:
> Currently, the session is not reset inside on_reset() since doing so would require another initialize() on the next call, adding more complexity.

Make sense.

ekzhu commented on python/packages/autogen-ext/src/autogen_ext/tools/mcp/_session.py:
```diff
@@ -26,3 +31,105 @@ async def create_mcp_server_session(
+
+    async def _run_actor(self) -> None:
+        try:
+            async with create_mcp_server_session(self.server_params) as session:
```
Maybe we can use `session.__aenter__()` and `session.__aexit__()` to manage the context manually?

SongChiYoung replied:
🧵 Why a queue-based actor
To be honest, I initially tried to directly use async with create_mcp_server_session(...) in various places — including inside adapters, tools, even agents. But I ran into all kinds of subtle and non-obvious issues:

RuntimeError: athrow(): asynchronous generator is already running

RuntimeError: Attempted to exit cancel scope in a different task than it was entered in

pytest hanging forever on await fut

__del__ or atexit failing silently

At first, I genuinely thought I was doing something wrong.
Turns out... I kind of was, but also not really. The root issue is that session lifecycle must be managed within a single, isolated task — especially since MCP servers use nested async generators and anyio.TaskGroup-based cancel scopes.

🧱 What this queue-based actor solves
This design spawns a dedicated _run_actor() task which:

Opens and manages the session lifecycle fully inside one task

Listens to commands via asyncio.Queue

Emits results via Future, supporting multi-tool access without race conditions

Shuts down gracefully — even when tests teardown early, or the user forgets to call close()

This solves all of the above edge cases, and makes it super easy to mock in tests.

🤔 About session.__aenter__() / __aexit__()
I originally tried this — even had code using them directly.
But it's really hard to get right in practice:

__aexit__() must run in the same task that opened the session

But Agent resets, tool sharing, or test teardown often destroy task-local assumptions

You’ll end up with generators that don’t stop, or cancel scopes that break the whole loop

✅ Summary
The queue-based actor may look a bit verbose, but it's the only approach I found that:

Works with multiple tools sharing one session

Doesn’t crash in pytest

Shuts down cleanly on atexit

Keeps session logic isolated, with no race conditions

Let me know if you'd like me to break it up into a smaller PR or rework it for more idiomatic Autogen component usage. Always happy to revise.

Thanks again 🙏 This fix really came out of trial, error, and lots of 😵‍💫.

ekzhu replied:
I think the single task based approach makes sense and it makes it easier to work with session.

ekzhu commented on python/packages/autogen-ext/src/autogen_ext/tools/mcp/_factory.py:
Let's update the API doc to use a context manager for the `McpSessionActor`.

ekzhu commented on python/packages/autogen-ext/src/autogen_ext/tools/mcp/_session.py:
```diff
@@ -26,3 +31,105 @@ async def create_mcp_server_session(
+    server_params: McpServerParams
+
+
+class McpSessionActor(ComponentBase[BaseModel], Component[McpSessionConfig]):
```
The `McpSessionActor` should be used as an async context manager.

it is very easy for user to accidently forget calling `close()`. 

I think the new usage should be like this:

```python
async McpSessionActor(server_params) as session:
  tools = mcp_server_tools(sessions)
  # rest of code.
```

Otherwise we will need to call `close()` on every tools directly, which is fine in some cases, but it is really easy to forget. So we should support the async context manager usage.

ekzhu reviewed (changes requested):
I have changed my mind regarding the overall design I approved earlier.

I tried this with some examples and I found that when there is an error from the tool, it is not properly closing the tools and causing event loop closed errors -- as I expected because I forgot to include try--finally to close the tools.

Try this simple example:

```python
    params = StdioServerParams(
        command="uvx",
        args=["mcp-server-fetch"],
        read_timeout_seconds=60,
    )
    tools = await mcp_server_tools(server_params=params)
    assert tools is not None
    assert tools[0].name == "fetch"
    try:
        result = await tools[0].run_json({"url": "https://github.com/microsoft/not_exist"}, CancellationToken())
        assert result is not None
    finally:
        await tools[0].close()
```

Another weird issue is when one of the tool is closed, all of the tools will be closed -- this can be counter-intuitive.

I think we still need to use a context manager for the session. We can still use the `McpSessionActor` wrapper. I think this way it will make the implementation much easier and avoid confusion.

One more note: I think `McpSession` instead of `McpSessionActor` is more succinct.

ekzhu commented:
> In most AutoGen use cases, we don’t call tools immediately after creation. Instead, tools are passed into Agents or Teams as dependencies.

Then all of the running of agents and teams should be within the session's context. I don't see why this is an issue. The session context clearly indicates the lifecycle of the session, and anything that depends on it should live within it.

If the session is not created as a context, then we can lazily initialize it. But how do you solve the issue that closing one tool closes all the tools?

SongChiYoung commented:
@ekzhu 
Thanks for the suggestion.  
I tried testing with a `with` block, but hit a few issues - so I wanted to first explain where things stand.

### TL;DR  
1. To fix #6198, all MCP tools must share the same session.  
    --> That’s why closing one tool closes all others.  
2. Using `with` solves that, but breaks user code of before version and makes serialization hard.  
   --> Found this thanks to your comment - turns out even my version has deserialization issues.  
3. The current Actor pattern avoids both problems.  
   --> Runs in its own task and owns the session.  
   --> Can inject the same session into multiple tools.

I’ll still try to support `with` in a clean way if possible.

---

<details>
<summary>1. #6198 root issue</summary>

The original issue was the session closing early if tools used `with` inside an Agent.

Later I found a bigger problem:  
**Tools in the same MCP server need to share the same session.**  

For example, `web_surf` and `save_pdf` are separate tools — but the PDF tool can’t access browsed pages unless they share a session.  
That’s why I had to inject the same session into all tools when using `mcp_server_tools()`.

</details>

<details>
<summary>2. Why `with` was tricky in practice</summary>

Yes, wrapping the whole thing in a `with` block technically works —  
But it breaks two things:

- It changes the public API and user code has to change
- It breaks serialization/deserialization  
  (Sessions are created outside the Agent/Team - so they can’t be deserialization)

Right now I’m working on making this session sharing work even in deserialization — your comment helped me catch this. Appreciate it.

</details>

<details>
<summary>3. Why I went with Actor</summary>

MCP SDK raises errors when the session is closed from a different task:

```
RuntimeError: athrow(): asynchronous generator is already running  
RuntimeError: Attempted to exit cancel scope in a different task than it was entered in
```

That’s why I put the session inside a dedicated task - the Actor.  
That way it’s always closed from the right place. Tools just talk to the Actor, and it owns the session.

</details>

<details>
<summary>4. About “closing one tool closes all”</summary>

This is expected - because tools share the same session.

I thought about tracking tool lifetimes and only closing the session when all are done -  
But it adds a lot of complexity, and probably isn’t worth it.  
If there’s a cleaner idea I’m open though.

</details>

Let me know if you think anything here’s off.  
Happy to adjust direction.

SongChiYoung commented:
@ekzhu 
I think the current implementation is now in a pretty good place, and I’d love your thoughts.

### ✅ Summary of changes:
- The session now lives inside `McpToolAdapter`, but **is not closed at the end of `with`** — this ensures it stays alive for as long as the Agent is running.
- I updated the `atexit` handler to avoid event loop closed errors — it should now exit more gracefully.
- I also fixed the issue where closing one tool would shut down the entire session — I'm now tracking a simple `ref_count` so the session only shuts down when all tools are done.
- Serialization and deserialization are working correctly again!

> One note on the `async with` usage pattern:  
> As I mentioned in an earlier comment, I’m hesitant to require users to wrap tools in a `with` block at a higher level — mainly because it introduces real complications with serialization/deserialization.  
> That’s why I’m keeping the session lifecycle managed independently for now.

### 🤔 A quick question about tests:
From what I saw in the previous `codecov` results, the coverage didn’t drop significantly after the changes.  
That said, if there are specific test cases you’d still like to see, I’d be happy to add them!  
Many of the new `raise` statements were added for safety, and might be tricky to trigger intentionally.

I’ve also documented more details, but figured I’d hold off unless it’s helpful now (since it might be too much information all at once). Just let me know if you’d like me to share more!

ekzhu commented:
>  It breaks serialization/deserialization
(Sessions are created outside the Agent/Team - so they can’t be deserialization)

I think the fundamental issue is that MCP server is not compatible with how `AssistantAgent` uses tools, which assumes tools are separate and stateless.

It's not something we can solve with this PR -- it will require changes to the `AssistantAgent`. We could pass the MCP server parameters as a constructor argument to `AssistantAgent`, which then setups the session internally and manages it as part of its own lifecycle. This way, it will work perfectly with serialization and deserialization.

SongChiYoung commented:
@ekzhu
You're absolutely right — I totally agree with your point.

But since passing `MCP server parameters` directly into `Agent` would introduce a breaking change,  
how about making `stateful tools` optional instead?

(That way, the current PR wouldn't need to change much at all.)

Also, we could add `close()` and `on_reset()` hooks to `Agent`:
- `close()` would terminate the session
- `on_reset()` would re-initialize it

BTW even with your design, I think we’ll still need the `Agent.close()` hook for cleaning up the session,  
and `on_reset()` must re-initialize it anyway.

so this might work either way.

ekzhu commented:
I think we can do that in a different PR. 

For this PR, I want to avoid the complexity in the current implementation: reference counts, actors ... too many moving parts and potential for bugs.

Let's just focus on addressing the original issue with minimal fix, without worrying about serialization, which will be addressed in a separate PR.

So, for tools that require sharing a persistent session:

```python
with create_mcp_server_session(server_params) as session:
  tools = mcp_server_tools(session)
  agent = AssistantAgent(..., tools=tools)
```

This won't work with serialization but that's not the point. It addresses the immediate issue.

For tools that can use new session on each invocation:

```python
tools = mcp_server_tools(server_params)
agent = AssistantAgent(..., tools=tools)
```

Same thing as before. 

I believe my issue description in #6198 already described this.

ekzhu commented:
I'd rather have a quick fix to this problem so I can make a release today :D 

I can take on this if it's too late for you.

SongChiYoung commented:
@ekzhu 
Totally fair — I understand the goal is to keep this PR minimal and focused. 👍

That said, I’ve been working on a use case where MCP + serialization is actually super important 😅
I’m building an internal tool CompanyDeepResearch on top of AutoGen + MCP, using the autogen-oaiapi wrapper I shared earlier in Discord.
For concurrent access and persistence, team (de)serialization for copy is kind of essential.

So… just wanted to flag that there’s at least one user (me 😅) really hoping to see serialization support without introducing breaking changes to the API.

I’ll hold off for now and follow up in a separate PR as suggested!

By the way — it’s currently 4AM KST here 😅  
I’m worried I might break something if I try to fix this right now — would you mind taking care of it for this issue? I’d really appreciate it! 🙏

ekzhu commented:
> That said, I’ve been working on a use case where MCP + serialization is actually super important 😅

Understood. Let's get this issue over with and work on a more complete fix as a separate PR.
